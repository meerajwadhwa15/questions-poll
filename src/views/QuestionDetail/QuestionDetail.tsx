import React, { Component } from "react";
import { connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";
import { Loader, NoRecordFound } from "@/components";
import { FetchQuestionActions } from "@/store/actions";
import Detail from "./components/Detail";

import {
  StateProps,
  ReducerTypeDetail,
  ReducerStateDetail,
  DispatchPropsDetail,
  ComponentProps
} from "@/types/questions";

export const mapStateToProps = ({
  QuestionsReducers: { questionDetail, isFetchingDetail }
}: ReducerTypeDetail): ReducerStateDetail => ({
    questionDetail,
    isFetchingDetail
  });

export const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchPropsDetail => ({
  fetchQuestion: (id: number) => {
    dispatch(FetchQuestionActions(id));
  }
});

type QuestionDetailPageProps = {
  match: {
    params: {
      id: number;
    };
  };
};

export type Props = ComponentProps & ReducerStateDetail & DispatchPropsDetail & QuestionDetailPageProps;

class QuestionDetailView extends Component<Props, StateProps> {
  componentDidMount() {
    const { fetchQuestion, questionDetail, match: { params: { id } } } = this.props;
    if (fetchQuestion && id && (!questionDetail || (questionDetail.id && questionDetail.id !== id))) {
      fetchQuestion(id);
    }
  }

  render() {
    const { questionDetail, isFetchingDetail } = this.props;
    return (
      <div>
        {isFetchingDetail ? (
          <Loader />
        ) : questionDetail ? (
          <Detail item={questionDetail} />
        ) : (
              <NoRecordFound>No Question Found</NoRecordFound>
            )}
      </div>
    );
  }
}

export { QuestionDetailView };
export default connect<
  ReducerStateDetail,
  DispatchPropsDetail,
  ComponentProps,
  ReducerTypeDetail
  >(
  mapStateToProps,
  mapDispatchToProps
  )(QuestionDetailView);
