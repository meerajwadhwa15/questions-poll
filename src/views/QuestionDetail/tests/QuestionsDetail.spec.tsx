import React from "react";
import { QuestionDetailView, mapStateToProps } from "../QuestionDetail";
import { create } from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

const questionDetail =
    {
        id: 1,
        question: "Favourite programming language?",
        published_at: "2020-03-12T12:42:40.310135+00:00",
        url: "/questions/1",
        choices: [
            {
                choice: "Swift",
                votes: 39,
                url: "/questions/1/choices/1"
            }
        ]
    };

const props = {
    questionDetail,
    isFetchingDetail: false,
    fetchQuestion: jest.fn(),
    match: {
        params: {
            id: 1
        }
    }
};

const propsWithNoDetail = {
    questionDetail: null,
    isFetchingDetail: false,
    fetchQuestion: jest.fn(),
    match: {
        params: {
            id: 1
        }
    }
};

describe("Questions Enzyme testing", () => {
    test("Should call fetchQuestions", () => {
        const spyFetchQuestionAction = jest.spyOn(propsWithNoDetail, "fetchQuestion");
        const component = create(
            <BrowserRouter>
                <QuestionDetailView {...propsWithNoDetail} />
            </BrowserRouter>
        );

        const instance = component.getInstance();

        instance.componentDidMount();
        expect(spyFetchQuestionAction).toBeCalled();
    });

    test("Should not call fetchQuestions", () => {
        const spyFetchQuestionAction = jest.spyOn(props, "fetchQuestion");
        const component = create(
            <BrowserRouter>
                <QuestionDetailView {...props} />
            </BrowserRouter>
        );

        const instance = component.getInstance();

        instance.componentDidMount();
        expect(spyFetchQuestionAction).not.toBeCalled();
    });

    test("Matches the snapshot with QuestionDetail", () => {
        const component = create(
            <BrowserRouter>
                <QuestionDetailView {...props} />
            </BrowserRouter>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test("Should test mapStateToProps", () => {
        expect(
            mapStateToProps({
                QuestionsReducers: { questionDetail, isFetchingDetail: false }
            })
        ).toEqual({
            questionDetail,
            isFetchingDetail: false
        });
    });
});
