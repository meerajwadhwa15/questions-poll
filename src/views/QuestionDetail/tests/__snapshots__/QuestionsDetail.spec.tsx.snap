// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions Enzyme testing Matches the snapshot with QuestionDetail 1`] = `
<div>
  <div
    className="question-detail"
  >
    <h1
      className="page-title"
    >
      Question: 
      <strong>
        Favourite programming language?
      </strong>
    </h1>
    <p>
      Published Date: 
      <strong>
        12/Jun/2020
      </strong>
    </p>
    <div
      className="choices"
    >
      <div
        className="choice"
      >
        <div>
          Choice: 
          <strong>
            Swift
          </strong>
        </div>
        <div>
          Votes: 
          <strong>
            39
          </strong>
        </div>
      </div>
    </div>
  </div>
</div>
`;
